File: app/src/main/java/net/emilla/chime/Custom.java
package net.emilla.chime;

import android.content.Context;
import android.content.SharedPreferences;
import android.media.MediaPlayer;
import android.net.Uri;

import net.emilla.config.SettingVals;

final class Custom implements Chimer {

    private final Uri[] mUris;

    /*internal*/ Custom(SharedPreferences prefs) {
        Chime[] chimes = Chime.values();
        int chimeCount = chimes.length;
        var uris = new Uri[chimeCount];

        for (int i = 0; i < chimeCount; ++i) {
            uris[i] = SettingVals.customChimeSoundUri(prefs, chimes[i]);
        }

        mUris = uris;
    }

    @Override
    public void chime(Context ctx, Chime chime) {
        Uri uri = mUris[chime.ordinal()];

        MediaPlayer player;
        if (uri != null) {
            player = MediaPlayer.create(ctx, uri);
            if (player == null) {
                // URI is broken
                player = Nebula.mediaPlayer(ctx, chime);
            }
        } else {
            player = Nebula.mediaPlayer(ctx, chime);
        }

        player.setOnCompletionListener(MediaPlayer::release);
        player.start();
    }

}

File: app/src/main/java/net/emilla/chime/Custom.kt
package net.emilla.chime

import android.content.Context
import android.content.SharedPreferences
import android.media.MediaPlayer
import android.net.Uri

/**
 * A custom-sound chimer built from user settings.
 *
 * @param ctx it's important to use application context to avoid memory leaks!
 * @param prefs used to fetch sound URIs from user settings.
 */
class Custom(private val ctx: Context, prefs: SharedPreferences) : Chimer {
    private val uris: Array<Uri?> = Chimer.customSounds(prefs)

    override fun chime(id: Byte) {
        var player = MediaPlayer.create(ctx, uris[id.toInt()])
            ?: MediaPlayer.create(ctx, Nebula.sound(id))
        // fall back to nebula URI is broken or null
        player.setOnCompletionListener(MediaPlayer::release)
        player.start()
    }
}
